x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/BaseRunAsSuperuserCommand.java

```diff
@@ -0,0 +1,281 @@
+ * filesystem write access to perform actions on the node as a superuser. It leverages temporary file realm users
+ * with a `superuser` role.
+ */
+public abstract class BaseRunAsSuperuserCommand extends KeyStoreAwareCommand {
+
+    private static final String[] ROLES = new String[] { "superuser" };
+    private static final int PASSWORD_LENGTH = 14;
+
+    private final Function<Environment, CommandLineHttpClient> clientFunction;
+    private final CheckedFunction<Environment, KeyStoreWrapper, Exception> keyStoreFunction;
+    private SecureString password;
+    private String username;
+    final SecureRandom secureRandom = new SecureRandom();
+
+    public BaseRunAsSuperuserCommand(
+        Function<Environment, CommandLineHttpClient> clientFunction,
+        CheckedFunction<Environment, KeyStoreWrapper, Exception> keyStoreFunction
+    ) {
+        super("description");
+        this.clientFunction = clientFunction;
+        this.keyStoreFunction = keyStoreFunction;
+    }
+
+    @Override
+    public void close() {
+        if (password != null) {
+            password.close();
+        }
+    }
+
+    @Override
+    protected final void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        validate(terminal, options, env);
+        KeyStoreWrapper keyStoreWrapper = keyStoreFunction.apply(env);
+        final Environment newEnv;
+        final Settings settings;
+        if (keyStoreWrapper != null) {
+            decryptKeyStore(keyStoreWrapper, terminal);
+            Settings.Builder settingsBuilder = Settings.builder();
+            settingsBuilder.put(env.settings(), true);
+            if (settingsBuilder.getSecureSettings() == null) {
+                settingsBuilder.setSecureSettings(keyStoreWrapper);
+            }
+            settings = settingsBuilder.build();
+            newEnv = new Environment(settings, env.configFile());
+        } else {
+            newEnv = env;
+            settings = env.settings();
+        }
+
+        ensureFileRealmEnabled(settings);
+        try {
+            final Hasher hasher = Hasher.resolve(XPackSettings.PASSWORD_HASHING_ALGORITHM.get(settings));
+            password = new SecureString(generatePassword(PASSWORD_LENGTH));
+            username = generateUsername();
+
+            final Path passwordFile = FileUserPasswdStore.resolveFile(newEnv);
+            final Path rolesFile = FileUserRolesStore.resolveFile(newEnv);
+            FileAttributesChecker attributesChecker = new FileAttributesChecker(passwordFile, rolesFile);
+
+            Map<String, char[]> users = FileUserPasswdStore.parseFile(passwordFile, null, settings);
+            if (users == null) {
+                throw new IllegalStateException("File realm configuration file [" + passwordFile + "] is missing");
+            }
+            users = new HashMap<>(users);
+            users.put(username, hasher.hash(password));
+            FileUserPasswdStore.writeFile(users, passwordFile);
```

Comment:
Do you think guaranteeing that the user has roles assigned when it exists is important?

To make it appear atomic, maybe it would be better to first update the roles file and then the users file (otherwise there are states where the user exists but doesn't have a role, and we ought to handle this case as a retry).
In addition, we would also have to change the server code to ensure that if the roles file is changed before the users file, changes to the roles file are always considered when updating the users.

Reply:
> Do you think guaranteeing that the user has roles assigned when it exists is important?

Yes!Thanks. 

> To make it appear atomic, maybe it would be better to first update the roles file and then the users file

I think that's a good idea nevertheless

> In addition, we would also have to change the server code to ensure that if the roles file is changed before the users file, changes to the roles file are always considered when updating the users.

Not sure I follow this, maybe we can quickly chat in our sync

Reply:
Changed the order of execution here and added 403 to the list of error codes we should retry upon